File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    if target not in words:
        return 0
    m = 51
    visited = [0 for x in range(len(words))]
    def dfs(word, target, val):
        nonlocal m
        if word == target:
            if m > val:
                m = val
        else:
            for i in range(len(words)):
                cnt = 0
                for j in range(len(word)):
                    if word[j] != words[i][j]:
                        cnt+=1
                if cnt == 1 and visited[i] == 0:
                    visited[i] = 1
                    dfs(words[i],target,val+1)
                    visited[i] = 0
    dfs(begin, target, 0)
    return m

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const canChange = (word1,word2) => {
        const w1 = word1.split("")
        const w2 = word2.split("")
        let cnt = 0
        for(let i=0;i<w1.length;i++){
            if(w1[i]!==w2[i]) cnt++;
            if(cnt>=2) return false
        }
        return true
    }
    let ans = 0
    let visited = Array(words.length).fill(false)
    const bfs = () => {
        let queue = [begin]
        while(queue.length>0){
            let len = queue.length
            for(let i=0;i<len;i++){
                let curr = queue.shift()
                if(curr===target) return ans
                for(let j=0;j<words.length;j++){                    
                    if(visited[j]) continue
                    if(canChange(curr,words[j])) {
                        if (words[j] === target) return ans + 1;
                        visited[j] = true
                        queue.push(words[j])
                    }
                }
            }
            ans++
        }
        return 0
    }
    return bfs()
}